### PR TITLE
new EXTCODEHASH tests

### PR DESCRIPTION
**Pull request by hugo-dc**
- [x] EXTCODEHASH to an account without code - https://github.com/ethereum/tests/issues/540
- [x] EXTCODEHASH to a non existing account - https://github.com/ethereum/tests/issues/541
- [x] EXTCODEHASH of contract with codesize equal to maxcodesize - 1 (24576 bytes)
- [x] EXTCODEHASH of precompiles accounts (1-8)

### CODE REVIEW

**Comment by holgerd77** [comment]
👍 👍 👍

**Comment by winsvega** [comment]
Dont cover all tests in one go. Make several small PRs

**Comment by winsvega** [comment]
Which testeth commit you use? 
Filler hash is either outdated testeth issue or the way you construct yaml sources.

Make sure the fields are string and int accordingly: https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stArgsZeroOneBalance/eqNonConstFiller.yml

**Comment by holgerd77** [comment]
Question for the docs: what is the current state of .yml vs .json? Are both formats supported equally? or is .yml to be preferred (and .json even to be removed in the future?)?

Then I would eventually update all the filler .json examples in the docs with the corresponding .yml variant.

**Comment by winsvega** [comment]
both are supported. no preference. 
the issue is that testeth does not check yml semantics. so you could put string instead of int and vice verca and it will still be able to run it. while this kind of change is crucial to hash calculation by the python script.

**Comment by holgerd77** [comment]
Ah, thanks!

**Comment by hugo-dc** [comment]
> Which testeth commit you use?

I tried with testeth 1.4.0, then tried with testeth 1.5.0.dev2-76+commit.2fd57400, both give different hashes and both fail in travis ci

> Filler hash is either outdated testeth issue or the way you construct yaml sources.

I'll check the yaml files to see if there is something to fix

File: src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml [hunk 1]
```diff
@@ -0,0 +1,160 @@
+# EXTCODEHASH of precompiles addresses
+---
+extCodeHashPrecompiles:
+  env:
+    currentCoinbase: 2adc25665018aa1fe0e6bc666dac8fc2697ff9ba
+    currentDifficulty: '0x20000'
```
Comment (hugo-dc) [inline src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml:6]: The problem was I wasn't using quotes for this value

**Comment by hugo-dc** [comment]
@winsvega @holgerd77 this is ready for review, I'll continue working on new tests in a different PR

**Comment by holgerd77** [comment]
@hugo-dc Great, thanks so much for your work Hugo! 👍

File: src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    aaaaaaaa00000000000000000000000000000002:
+      balance: '1000000000000000000'
+      code: |
+        {
+          (MSTORE 0 (EXTCODESIZE 0x0000000000000000000000000000000000000002))
+          (RETURN 0 32)
+        }
+      nonce: '0'
+      storage: {}
+    # account with random bytecode with code size = MAXCODESIZE - 1 = 24575
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml]: comment from the previous test file. no longer actual

File: src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        4: "0x01",
+        5: "0x01",
+        6: "0x01",
+        7: "0x01"
+        }
+    aaaaaaaa00000000000000000000000000000001:
+      balance: '1000000000000000000'
+      code: |
+        {
+          (MSTORE 0 (EXTCODEHASH 0x0000000000000000000000000000000000000001))
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml]: extcodehash / extcodsize  are mixed here (in this test) inconsistent. 
you could also pass the contract that you want to call as DATA for the call.

```
(MSTORE 0 0x1122...32) (CALL 50 account 0 0 32 0 0)
```

and load that data in account with `(CALLDATALOAD 0)`
so you would have one contract as function that execute and store return values of extcodesize and extcodehash  of an input argument. then you just use different data in general transaction to call that contract/function with different data.  in this case the data (argument for the function) is precompiled addresses.

File: src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml (hunk 1 above)
Comment (winsvega) [reply src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml:6]: please use the latest master branch testeth as it must work anyway.

**Comment by winsvega** [comment]
related: https://github.com/ethereum/tests/issues/541
related: https://github.com/ethereum/tests/issues/540
require testeth update if that wasnt the case. 

(I mean if those conditions are not true with the latest testeth then cpp aleth must be fixed and testeth must be rebuilt)

File: src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml (hunk 3 above)
Comment (hugo-dc) [reply src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml]: Thanks for reviewing, I have updated the test case according to your comments. I'm sending the precompiles addresses as data. I updated the test cases using latest aleth/testeth master (i had to change it to accept the stExtCodeHash directory), I don't know if that is what you meant or if I should use https://github.com/ethereum/testeth instead.

File: GeneralStateTests/stExtCodeHash/extCodeHashAccountWithoutCode.json [hunk 4]
```diff
@@ -0,0 +1,84 @@
+{
+    "extCodeHashAccountWithoutCode" : {
+        "_info" : {
+            "comment" : "",
+            "filledwith" : "testeth 1.5.0-alpha.6-1+commit.19ad7d95.dirty",
```
Comment (winsvega) [inline GeneralStateTests/stExtCodeHash/extCodeHashAccountWithoutCode.json]: .dirty means you have uncommited changes to testeth. thats why. 
ideally a PR to aleth repo is open. testeth modified with new test suites and commited. (so no dirty flag) 
then aleth run it on a specific tests branch with new tests. (also set in a PR to aleth) 

*note you dont need to do that logic. as it is already done on aleth create2 branch. what I meant was 
if testeth from create2 branch or the latest version of testeth that you use produce incorrect test results (from one expected by EIPs) then testeth must be updated (and aleth as well) 

https://github.com/ethereum/tests/issues/540
https://github.com/ethereum/tests/issues/541

File: src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml (hunk 1 above)
Comment (winsvega) [reply src/GeneralStateTestsFiller/stExtCodeHash/extCodeHashPrecompilesFiller.yml:6]: oh. this comment was about failing test hash issues. but seems like hash reports from python script are due to the .yml files format. (quotes)

**Comment by winsvega** [comment]
looks good. 
the .yml format though keep failing by python script check. (which is done separate from testeth) thats why you could not check it while filling.
and ideally would be to get rid of `.dirty` flag in testeth commit version.

**Comment by hugo-dc** [comment]
@winsvega i refilled tests using the `create2` branch of aleth, now all validation tests are passing.

**Comment by winsvega** [comment]
looks good. please run --filltests --fillchain and commit the changes

**Comment by hugo-dc** [comment]
filled blockchain files

**Comment by holgerd77** [comment]
Hi @hugo-dc, congrats, super cool, I actually wasn't aware that you have got such deep knowledge on test creation! 😄

One thing: these additional "info" fields in the fillers are actually super-important for using the tests on the client side, people are otherwise spending up to hours figuring out what a test case really is about (I know there is some hint through the test case naming, but this is normally not detailed enough).

I know this is a bit annoying but could you do this extra round and add info-comments to the ``yml`` files? That would be great! There also (probably?) shouldn't be the need to re-fill or something, this will organically happen somewhat later, so just a PR against the initial yml fillers.

[Here](https://github.com/ethereum/tests/blob/428842e9731b7065366dcd7829bbfc0f9e1135fa/src/GeneralStateTestsFiller/stSystemOperationsTest/doubleSelfdestructTestFiller.yml) is an example of a comment integrated in a yml filler (just for the syntax, write in whatever you like).

**Comment by holgerd77** [comment]
Additional note: have just seen that you have added explanations as comments in the beginning of the yml files. This would nevertheless be better to be placed in the info comment fields, since this is then also transported to the filled tests and can there further be used (e.g. also programatically by printing out when running the tests).

**Comment by winsvega** [comment]
To understand what test does you better study filler.json or filler.yml which has comments and source code of the bytecode

**Comment by hugo-dc** [comment]
@holgerd77 thanks for your comments!, definitely it would be better to include detailed comments to the `_info` field, I will consider it for the next test cases, I can add it to these tests as well, thanks for showing an example.

**Pull request merged**